### PR TITLE
many: struct-based revisions, new representation

### PR DESCRIPTION
**Pull request by niemeyer**
This bases on #1164 and:

  * Moves the snap.Revision type to be a struct
  * Adds Local/Store/Unset helper methods on it
  * Changes its String so negative numbers are "x%d"
  * Rollbacks the API to int-based revisions
  * Introduces a richer Revision type in the client

### CODE REVIEW

**Comment by niemeyer** [comment]
This is ready for a couple of reviews.

Please don't get scared by the branch size. Most of the changes are repetitive and trivial.

File: client/revision_test.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+func (s revisionSuite) ParseRevision(c *C) {
+	type testItem struct {
+		s string
+		n int
+		e string
+	}
+
+	var tests = []testItem{{
```
Comment (mvo5) [inline client/revision_test.go]: Nice!

File: integration-tests/tests/list_test.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ func (s *listSuite) TestListMustPrintAppVersion(c *check.C) {
 	expected := "(?ms)" +
 		"Name +Version +Rev +Developer *\n" +
 		".*" +
-		"^hello-world +(\\d+)(\\.\\d+)* +[0-9]+ +.* *\n" +
+		"^hello-world +(\\d+)(\\.\\d+)* +[x0-9]+ +.* *\n" +
```
Comment (mvo5) [inline integration-tests/tests/list_test.go]: In this particular case the snap comes from the store so we know that there will be no `x` in here. However its not the main purpose of the test and certainly not a blocker if it gets added.

File: snappy/overlord.go [hunk 3]
```diff
@@ -438,7 +438,7 @@ func (o *Overlord) InstallWithSideInfo(snapFilePath string, sideInfo *snap.SideI
 	// XXX: this is still done for now for this legacy Install to
 	// keep unit tests as they are working and as strawman
 	// behavior for current u-d-f
-	if newInfo.Revision != 0 { // not sideloaded
```
Comment (mvo5) [inline snappy/overlord.go]: Good catch!

**Comment by mvo5** [comment]
Looks good, :+1:

File: integration-tests/tests/list_test.go [hunk 4]
```diff
@@ -44,7 +44,7 @@ func (s *listSuite) TestListMustPrintCoreVersion(c *check.C) {
 	expected := "(?ms)" +
 		"Name +Version +Rev +Developer *\n" +
 		".*" +
-		fmt.Sprintf("^%s +.* +%s + [0-9]+ +(canonical|sideload) *\n", partition.OSSnapName(c), verRegexp) +
+		fmt.Sprintf("^%s +.* +%s + [x0-9]+ +(canonical|sideload) *\n", partition.OSSnapName(c), verRegexp) +
```
Comment (chipaca) [inline integration-tests/tests/list_test.go]: it's not `[x0-9]+`, it's `x?[0-9]+`

Comment (chipaca) [inline integration-tests/tests/list_test.go]: actually `x?[1-9][0-9]*` if you want to be really nitpicky

File: daemon/api.go [hunk 5]
```diff
@@ -382,7 +382,7 @@ func searchStore(c *Command, r *http.Request, user *auth.UserState) Response {
 	for i, x := range found {
 		url, err := route.URL("name", x.Name())
 		if err != nil {
-			logger.Noticef("cannot build URL for snap %q (r%d): %v", x.Name(), x.Revision, err)
+			logger.Noticef("cannot build URL for snap %q revision %s: %v", x.Name(), x.Revision, err)
```
Comment (chipaca) [inline daemon/api.go]: while you're here, could you capitalize the cannot? given it's a noticef. ditto further down.

**Comment by chipaca** [comment]
OK. Starting to wonder whether this still counts as "just make it an int"... :-)
:+1:

File: integration-tests/tests/list_test.go (hunk 2 above)
Comment (niemeyer) [inline integration-tests/tests/list_test.go]: Removed the x, thanks.

File: integration-tests/tests/list_test.go (hunk 4 above)
Comment (niemeyer) [inline integration-tests/tests/list_test.go]: Remove the x altogether here too. It also comes from the store, so there's no need.

Also removed the "sideload" message from the right hand side.

File: daemon/api.go (hunk 5 above)
Comment (niemeyer) [inline daemon/api.go]: Done, thanks.

**Pull request merged**